### PR TITLE
Add "compression" option to otlphttp exporter

### PR DESCRIPTION
**Pull request by tigrannajaryan**
The default value of "compression" option is empty and results in no compression,
which is the same as the old behavior. You can also specify "compression: gzip"
which will result in gzip encoding of outgoing http requests.

### CODE REVIEW

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:49]: We should be consistent with gRPC config

Comment (tigrannajaryan) [reply config/confighttp/confighttp.go]: Fixed.

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:46]: There is a problem here. It is not guaranteed that multiple goroutines are not calling this at the same time. So probably allocate a new zipper or use a pool

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:55]: I am confused why you close this but the received one was not closed.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go:46]: Let me double check. I believe this is one roundtripper per http client and you are not allowed concurrent calls on the same http client.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go:55]: Do you mean the response? The response is returned to the caller, they are responsible for closing.

File: internal/middleware/compression.go
Comment (bogdandrutu) [reply internal/middleware/compression.go:46]: That can only be possible if the http client serializes all the requests

File: internal/middleware/compression.go
Comment (bogdandrutu) [reply internal/middleware/compression.go:55]: I mean the 4 lines above. Body doesn't seem to be closed when you receive it. You either do this and not close the body or close the body but not this.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go:46]: OK, I can't find anything in the docs, so I think you are right since it is not prohibited. Let me fix.

Comment (tigrannajaryan) [reply internal/middleware/compression.go]: Changed to use a new zipper on every request. 
Not sure about the performance though, I don't know how costly it is to recreate gzipWriter on every request. If it shows to be slow may need to add a pool of gzip writers to use instead. I am keeping simple for now, can be improved later.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go]: Fixed.

**Comment by tigrannajaryan** [comment]
Don't merge yet, I need to check all contrib exporters to make sure they are not affected.

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:46]:
```golang
buf := bytes.NewBuffer([]byte{}) // Maybe preallocate this with WriteBufferSize? Just a TODO :)
gzipWriter := gzip.NewWriter(buf)
```

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:48]: Check for error?

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:66]: Based on other places where we do in-place changes I think you should not worry to clone https://github.com/open-telemetry/opentelemetry-collector/blob/846b971758c92b833a9efaf742ec5b3e2fbd0c89/config/confighttp/confighttp.go#L91

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:58]: Link?

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go:72]: The docs say:
```
// RoundTrip should not modify the request, except for
// consuming and closing the Request's Body.
```

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go]: Added doc link.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go:52]: Not sure, may be premature optimization. I suggest to keep as is and can be changed if profiling shows it is needed.

**Comment by tigrannajaryan** [comment]
OK, changed the implementation to make "compression" option opt-in for exporters. It was dangerous to enable for all http exporters since some exporter already do their own compression. See updated PR description.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go]: Done.

**Comment by tigrannajaryan** [comment]
contrib_tests failure is expected. Will need to fix after this is merged.

**Comment by tigrannajaryan** [comment]
Now ready for review.

File: exporter/otlphttpexporter/otlp_test.go
Comment (jpkrohling) [inline exporter/otlphttpexporter/otlp_test.go:172]: This is technically an `assert`

File: internal/middleware/compression_test.go
Comment (jpkrohling) [inline internal/middleware/compression_test.go:66]: Anything against https://golang.org/pkg/net/http/httptest/#Server ? Looks more suitable for this case here.

Comment (tigrannajaryan) [reply internal/middleware/compression_test.go:66]: Nothing against :-) I just reused the approach used by the existing test in this same file. Can be refactored in a separate PR.

File: exporter/otlphttpexporter/otlp_test.go
Comment (tigrannajaryan) [reply exporter/otlphttpexporter/otlp_test.go]: Fixed.

**Review by jpkrohling — APPROVED** [review]

File: internal/middleware/compression.go
Comment (bogdandrutu) [reply internal/middleware/compression.go:54]: The main suggestion not the extra comment is not an optimization, but more of a style change. Initialization of the zipper with the buffer instead of init with nil then reset

**Review by bogdandrutu — COMMENTED** [review]
Personally I think this is too much of a change for 3-4 lines directly in the exporter. I would suggest that we don't implement a Roundtrip and do directly this in the exporter code and add a `compression` property there.

This change will be bad because will expose a `compression` for all the configs but some will not enable it.

**Comment by tigrannajaryan** [comment]
> Personally I think this is too much of a change for 3-4 lines directly in the exporter. I would suggest that we don't implement a Roundtrip and do directly this in the exporter code and add a `compression` property there.
> 
> This change will be bad because will expose a `compression` for all the configs but some will not enable it.

@bogdandrutu This implementation is pretty uniform with how GRPCClientSettings sees compression. I think it will be useful to add "compression" to other exporters (e.g. Zipkin) with just one line change.

If you think don't think that other exporters need it then I can do what you suggest and just have the compression implemented for otlphttpexporter only.

Your call, I don't have a strong opinion.

**Comment by bogdandrutu** [comment]
> Your call, I don't have a strong opinion.

I would vote to make it only for this exporter but reuse the same name for configuration.

File: internal/middleware/compression.go
Comment (tigrannajaryan) [reply internal/middleware/compression.go]: Ah, I did not realize what you mean. Done.

**Comment by tigrannajaryan** [comment]
> > Your call, I don't have a strong opinion.
> 
> I would vote to make it only for this exporter but reuse the same name for configuration. The configuration property will be defined in the exporter Config at the moment.

@bogdandrutu done. PTAL.

File: internal/middleware/compression_test.go
Comment (tigrannajaryan) [reply internal/middleware/compression_test.go:66]: I assume we are good here, resolving. Please reopen if needed.

File: testbed/tests/trace_test.go
Comment (bogdandrutu) [inline testbed/tests/trace_test.go:118]: What about we add the gzip to the already existing `TestTrace10kSPS` to avoid running the non-gzip twice?

File: internal/middleware/compression.go
Comment (bogdandrutu) [inline internal/middleware/compression.go:34]: I will give a try after this PR merges to just do this directly in the exporter and see if we save some performance or code gets smaller.

Comment (tigrannajaryan) [reply internal/middleware/compression.go:34]: SGTM.

Comment (tigrannajaryan) [reply internal/middleware/compression.go:34]: It would be interesting to profile and see how much of a difference it makes (assuming we don't resintantiate gzipwriter on every request). It may be significant for very small requests (e.g. one span requests).

File: testbed/tests/trace_test.go
Comment (tigrannajaryan) [reply testbed/tests/trace_test.go]: Done.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
